### ISSUE TITLE
Touch events not converted to full clicks

### ISSUE THREAD

**Issue by DDeas** [cefsharp/CefSharp#1326]
OK, so the newer public release 43.0.0 contains a workaround to circumvent an issue causing sticky context menus.

Unfortunately, this fix (OnIsBrowserInitializedChanged override) has introduced another UI bug.

The fix essentially simulates a mouse click to the browser controls containing form, whenever it receives a mouse activate message (so the first click that focuses/activates the control). This control activating mouse click in my case has been converted from a screen touch.

If you try and touch a form element (so menus, buttons inside docked panels etc) external to the control after that first click/touch, the element will respond as normal. If however, you have clicked/touched inside the browser control more than twice, the external element will no longer respond to your touches. You have to click with a physical mouse to restore touch events to external controls.

Introduced in 43.0.0. This also occurs in the latest master branch (last pulled on Friday).

**Comment by amaitland** [comment]
> OK, so the newer public release 43.0.0 contains a workaround to circumvent an issue causing sticky context menus.

What section of code are you referring to explicitly?

**Comment by DDeas** [comment]
```csharp

public BrowserTabUserControl(string url)
        {
            var browser = new ChromiumWebBrowser(url)
            {
                Dock = DockStyle.Fill
            };
            browser.IsBrowserInitializedChanged += OnIsBrowserInitializedChanged;
        }

private void OnIsBrowserInitializedChanged(object sender, IsBrowserInitializedChangedEventArgs args)
        {
            if (args.IsBrowserInitialized)
            {
                ChromeWidgetMessageInterceptor.SetupLoop((ChromiumWebBrowser)Browser, (message) =>
                {
                    const int WM_MOUSEACTIVATE = 0x0021;
                    const int WM_NCLBUTTONDOWN = 0x00A1;

                    if (message.Msg == WM_MOUSEACTIVATE) {
                        // The default processing of WM_MOUSEACTIVATE results in MA_NOACTIVATE,
                        // and the subsequent mouse click is eaten by Chrome.
                        // This means any .NET ToolStrip or ContextMenuStrip does not get closed.
                        // By posting a WM_NCLBUTTONDOWN message to a harmless co-ordinate of the
                        // top-level window, we rely on the ToolStripManager's message handling
                        // to close any open dropdowns:
                        // http://referencesource.microsoft.com/#System.Windows.Forms/winforms/Managed/System/WinForms/ToolStripManager.cs,1249
                        var topLevelWindowHandle = message.WParam;
                        PostMessage(topLevelWindowHandle, WM_NCLBUTTONDOWN, IntPtr.Zero, IntPtr.Zero);
                    }

                    // The ChromiumWebBrowserControl does not fire MouseEnter/Move/Leave events, because Chromium handles these.
                    // However we can hook into Chromium's messaging window to receive the events.
                    //
                    //const int WM_MOUSEMOVE = 0x0200;
                    //const int WM_MOUSELEAVE = 0x02A3;
                    //
                    //switch (message.Msg) {
                    //    case WM_MOUSEMOVE:
                    //        Console.WriteLine("WM_MOUSEMOVE");
                    //        break;
                    //    case WM_MOUSELEAVE:
                    //        Console.WriteLine("WM_MOUSELEAVE");
                    //        break;
                    //}
                });
            }
        }
```

**Comment by amaitland** [comment]
> OK, so the newer public release 43.0.0 contains a workaround to circumvent an issue causing sticky context menus.

That's only in the example project, not part of `CefSharp` it's self.

**Comment by DDeas** [comment]
OK, but its code we have to implement in order to use cefsharp, as has been presented to us as part of the final solution.

I'm hoping that this discussion will help build my understanding of whats happening around touches/mouseclicks/focuses etc, as I have to now come up with a re-circumvention to a circumvention to a problem I wasn't event having, that accidentally fixed a problem I was.

Why when a form element receives a simulated mouse click in this way, does further touch events into your cefsharp webbrowser control then cause the form to stop seeing touch events?

**Comment by amaitland** [comment]
> OK, but its code we have to implement in order to use cefsharp, as has been presented to us as part of the final solution.

It's just one users workaround for a problem they were having, there for others to use/draw from. 

> Why when a form element receives a simulated mouse click in this way, does further touch events into your cefsharp webbrowser control then cause the form to stop seeing touch events?

That I don't know. Your likely better of removing the workaround and coming up with another solution. Is it just the menu you need to hide?

Personally I'd see if the problem reproduced in the `cefclient` example application, if it does then post on `ceforum` for help, if it doesn't then look at it's source code.

Remember this is an open source project, not a commercial application with paid support, the project is only as good as the amount of time/effort people contribute!

**Comment by DDeas** [comment]
> Remember this is an open source project, not a commercial application with paid support, the project is only as good as the amount of time/effort people contribute!

Of course, I totally appreciate that, and am in awe of the achievements you guys have collectively made towards this. I was just explaining my position is all.

The circumvention is the only solution I have found to the stack overflow issue cause by typing over selected text on touch screen devices https://github.com/cefsharp/CefSharp/issues/1316 and http://magpcss.org/ceforum/viewtopic.php?f=6&t=13472&e=0

Ill take a look at the cefclient. I felt it was .Net control issue though, but I can have a go at stepping through the code myself.

Ill leave the call open to post results. Thanks for your time.

**Comment by DDeas** [comment]
Issue in more detail:

http://stackoverflow.com/questions/33104877/c-sharp-windows-touch-click-focus-messages-between-form-and-controls-wm-xx

**Comment by amaitland** [comment]
> OK, so the newer public release 43.0.0 contains a workaround to circumvent an issue causing sticky context menus.

I may have a different solution for this. It appears the problem is that by default `CefSharp` uses the `MultiThreadedMessageLoop` option, which spawns off a separate thread for the `CEF` message loop. Switching to calling `CefDoMessageLoopWork ` on `Application.Idle` means all messaging occurs in the same thread. Quick test shows this resolves the need for using the workaround to post click events. I haven't done more than a preliminary test.

`master` has been updated to expose `Cef.DoMessageLoopWork()`. I've added a rough example to the `WinForms.Example` project. It's effectively disabled by default.
https://github.com/cefsharp/CefSharp/commit/fe11f2eada542f49e4eef0feed9b2b0978446bbf#diff-f142d024925e73816b0fabd620d2f71dR37

@jankurianski This may be of some interest to you.
